Pulse queue size card

lehecht opened the pull request:
Add a pulse card to display the queue sizes.

Closes [biigle\biigle.de#117](https://github.com/biigle/biigle.de/issues/117)

lehecht committed:
Initial commit

lehecht committed:
Update composer

lehecht committed:
Add card view

lehecht committed:
Add livewire view controller

lehecht committed:
Minor changes in pulse card view

lehecht committed:
Add queue size recorder

lehecht committed:
Remove log

lehecht committed:
Add service provider

lehecht committed:
Record only aggregated values

lehecht committed:
Use Queue::size instead of counting events

lehecht committed:
Add tests

lehecht committed:
Enable lazy loading

lehecht committed:
Minor changes

lehecht committed:
Rename column

lehecht committed:
Remove duplicated line

lehecht committed:
Edit queue size card view

lehecht committed:
Add readme

lehecht committed:
Remove gitignore

lehecht committed:
Add infos about the project to the composer file

lehecht committed:
Remove pulse requirement

lehecht committed:
Remove composer lock file

lehecht committed:
WIP show queue size for all queues

lehecht committed:
Show chart for each queue

lehecht committed:
Save all queue types in cache

lehecht committed:
WIP: Display queues with aligned dates

lehecht committed:
Listen only to SharedBeat event

Record queue sizes only every 60s

lehecht committed:
Show graphs for each queue

lehecht committed:
Display all queues in a single canvas

lehecht committed:
Use only queue name

lehecht committed:
Publish config when pulse starts

lehecht committed:
Update queue size card when period is changed

lehecht committed:
Remove unnecessary method

lehecht committed:
Remove unused parameter

lehecht committed:
Improve queue color assignment

lehecht committed:
Update chart when new queues are available

lehecht committed:
Minro changes

lehecht committed:
Fix return value

lehecht committed:
Edit queue names

lehecht committed:
Update config

lehecht committed:
Update tests

lehecht committed:
Format code

lehecht committed:
Include default queue as default value

lehecht committed:
Fix new lines

lehecht committed:
Add filter for entry key

lehecht committed:
Add tests for controller

lehecht committed:
Update README

lehecht committed:
Update README

lehecht committed:
Remove comment

mzur reviewed (changes requested):
High-level comments first:

- I have published this at Packagist. It can now be installed with:
   ```
   composer require biigle/pulse-queue-size-card:dev-pulse-queue-size-card
   ```

- The `laravel/pulse` dependency is missing from the composer.json (or at least `livewire/livewire`).

- The pulse-ext config is not initialized correctly. This must be added to the register method of the service provider:
    ```php
    $this->mergeConfigFrom(__DIR__.'/config/pulse-ext.php', 'pulse-ext');
    ```

- Queues to monitor should have an optional queue connection specified. We can use the same syntax than the queue:monitor command: `my_queue` (uses the default connection), `my_connection:my_queue` (uses a different connection).
 
- The `pulse:check` command runs on many machines. Does the queue size recorder run only once or once for every machine (we don't want that)?
 
- There is one line per monitored queue in the chart (see below). I'd prefer one chart per queue as with the "queues" recorder. In each chart you could show three lines (pending, delayed, reserved) as the queue:monitor command does. Check the command to see how to get the numbers.
    <img width="1283" height="466" alt="Image" src="https://github.com/user-attachments/assets/f8a032d9-4107-48be-bfc9-3467fbf3f057" />
- There is no y-axis indicator in the chart (see above).

- Does Pulse clean up records older than 24h or are they kept indefinitely?

- The readme should say that the package requires a running `pulse:check` process.

lehecht committed:
Add laravel pulse dependency

lehecht committed:
Add mergeConfigFrom statement to provider

lehecht committed:
Update README

lehecht commented:
@mzur 
> * Does Pulse clean up records older than 24h or are they kept indefinitely?

According to [this comment](https://github.com/laravel/pulse/issues/294#issuecomment-1911235344), the records are deleted if they are not used for any time period. By default, this means if they are older than a week they will be deleted. Should I change this to 24 hours only for this feature or for all cards?

lehecht committed:
Display each queue in a separate canvas

lehecht committed:
Make adding of the queue connection optional

lehecht committed:
WIP: display queue states

lehecht committed:
Add queue size table

Pulse::set can save strings but overrides records,
while Pulse::record can only save numbers.

lehecht committed:
Use queue_size table

lehecht committed:
Create datasets from json objects

lehecht committed:
Ensure only a single machine uses the recoder

lehecht committed:
Update config

lehecht committed:
Delete records after 24h

lehecht committed:
Add test

lehecht committed:
Update config

lehecht committed:
Use config variable

lehecht committed:
Use migration with a model for better understanding

lehecht committed:
Add factory

lehecht committed:
Enable setting of cache lock key

lehecht committed:
Update tests

lehecht committed:
Add prune_after to config

lehecht committed:
Format code

lehecht committed:
Add phpdoc

lehecht committed:
Rename command for more clarity

lehecht committed:
Fix error with spacing

lehecht committed:
Revert to ff09de9

No separate scheduler needed since pulse tables should be used.

lehecht committed:
Partially revert to ff09de9

Save each queue value as record

lehecht committed:
Compute sum in view

lehecht committed:
Minor change

lehecht committed:
Update tests

lehecht committed:
Minor changes

lehecht committed:
Update config

lehecht committed:
Add new line

lehecht committed:
Format code

lehecht committed:
Edit phpdoc

lehecht committed:
Remove unused import

lehecht committed:
Simplify code

lehecht committed:
Add new line

lehecht committed:
Add filter to query and tests

lehecht committed:
Fix bug with missing queue states

Queue states are updated for each queue.
This can lead to missing states when
the states were changed for a certain period of time.

lehecht committed:
Update tests

lehecht committed:
Add comment

lehecht committed:
Edit comments

mzur commented on src/Http/Livewire/QueueSize.php:
```diff
@@ -16,46 +16,57 @@ public function render()
-            );
+            foreach ($queues->keys() as $key) {
+                $this->dispatch('queues-sizes-chart-update', queues: [$key => $queues[$key]]);
+            }
```
This looks like it could be done with a single update, too:
```
['key1' => [...], 'key2' => [...]]
```

mzur commented on resources/views/queue-size.blade.php:
```diff
@@ -2,13 +2,12 @@
     </x-pulse::card-header>
 
-    <div class="flex flex-wrap gap-4">
+    <div class="flex flex-wrap gap-4" style="margin-bottom: 15px;">
```
Prefer to put this into the card-header so you don't have to add a manual margin. Anyway, with Tailwind you would add the `mb-3` class instead of adding a manual style.

mzur commented on src/Recorders/QueueSize.php:
```diff
@@ -22,18 +49,39 @@ class QueueSize
      * Record the job.
      */
-    public function record(SharedBeat $event): void
+    public function record(): void
```
There is also an `IsolatedBeat` which could maybe be used instead of a manual lock? It runs every second, though.

lehecht replied:
@mzur 
> The pulse:check command runs on many machines. Does the queue size recorder run only once or once for every machine (we don't want that)?

Maybe I'm missing something, but then each machine running the `pulse:check` command would still execute the recorder. Or did you mean something different with your earlier comment?

mzur replied:
The way I understood `IsolatedBeat` was that it runs only on one machine, because it uses a lock (i.e. only one of the machines running `pulse:check` can get this lock). However, the event is sent every second so you may have to use additional logic to check the queue size only once a minute. If you had a look at this and found that `IsolatedBeat` works differently or you think it impractical then it's fine.

mzur commented on src/config/pulse-ext.php:
```diff
@@ -4,8 +4,11 @@
 
-    'queue_list' => 'pulse_queues',
+    // Must be a key from the queue:monitor output pointing to an integer
+    'queue_status' => ['pending', 'delayed', 'reserved'],
```
The comment seems to be misleading here. Also I don't see a need for this to be configurable. The different "sub"-queues are always called like this in Laravel.

mzur reviewed (changes requested):
The x-axis scaling of the chart seems to be wrong. In the screenshot below I submitted 4 jobs in total. The "queues" chart at the left shows when it happened on the correct scale. In "queue sizes" the scale is different/wrong:

<img width="1288" height="463" alt="Image" src="https://github.com/user-attachments/assets/2138fe12-6e1c-4522-8d3e-1cf509268c08" />

When I change the time interval at the top right (1h, 6h, 24h, etc) the chart does not update the x-axis, too.

Also the y-axis does not match. It correctly shows 4 in the purple indicator but the line is not at the top of the chart (where the indicator points at).

When there is no job in the queue, the indicator shows 0 although the line in the chart is still at 4:

<img width="1288" height="463" alt="Image" src="https://github.com/user-attachments/assets/06551cdb-6f93-4138-84b1-c506fd6c0e23" />

Could the configuration also be done in the pulse recorder definition instead of a separate configuration file? Example:
```php
Biigle\PulseQueueSizeCard\Recorders\QueueSize::class => [
   'enabled' => env('PULSE_QUEUE_SIZE_ENABLED', true),
   'sample_rate' => env('PULSE_QUEUES_SAMPLE_RATE', 1),
   'queues' => ['default', 'gpu:default', '...'],
   'record_interval' => 60,
],
```

Looking at this... is the `sample_rate` even used here? Or can it be used instead of `record_interval`?

lehecht commented:
@mzur 
> When there is no job in the queue, the indicator shows 0 although the line in the chart is still at 4:

Can you explain a bit more what you mean? As I understand, the queue sizes should be displayed over time. So at the right of the canvas, the line goes down to 0 since the queue is empty. Or did you mean the line is still at 4 in comparison to the regular queue card?

mzur commented:
The line is fine (except of the mentioned x and y axis scaling). I'm talking about the number in the purple indicator. If the current maximum of the line is at 4, the indicator should show 4 as well, although right now no jobs are in the queue any more.

lehecht commented:
@mzur 
Ah, so you mean the number in the badge should display the current maximum of all values, and not the current queue size as it does now?

mzur commented:
Right, it is the y-axis label.

lehecht committed:
Fix queue badge count

Badge count represents max value

lehecht commented:
@mzur 
> Looking at this... is the `sample_rate` even used here? Or can it be used instead of `record_interval`?

It can be omitted as it seems that 1 is the default value anyway. The [sample_rate](https://laravel.com/docs/12.x/pulse#sampling) gives you the option to control how many events should be recorded (e.g., 1 => all events, 0.3 => ~30% of events, etc.). Please tell me if you prefer the usage of the `sample_rate` over a record interval.

lehecht committed:
Use isolatedBeat instead of cache

lehecht commented:
@mzur 
When I move my config variables to the pulse config, I no longer have access to those variables during testing. However, Laravel recommends to use the [Orchestral Testbench](https://laravel.com/docs/12.x/packages#a-note-on-facades) for packages. Then it would be possible again. How should I proceed?

mzur commented:
We can keep using `record_interval`.

Can't you set (and use) `config('pulse.recorders.'.QueueSize::class)` in the tests?

lehecht committed:
Change axis scaling

lehecht committed:
Merge pulse-ext config with pulse config

lehecht committed:
Fix config var names

lehecht committed:
Update tests

lehecht committed:
Fix config values

lehecht committed:
Simplify code

lehecht committed:
Update tests

lehecht committed:
Update README

lehecht committed:
Use graph to retrieve data

No need for custom method since this card should work like the queue card.

lehecht committed:
Remove unused config file

lehecht committed:
Update tests

lehecht committed:
Remove unused sample rate

lehecht committed:
Update test

lehecht committed:
Use sum instead of average

lehecht committed:
Update tests

lehecht committed:
Update readme

lehecht committed:
Remove unused method

lehecht committed:
Improve code

lehecht committed:
Minor changes

lehecht committed:
Rephrase sentence in Readme

lehecht committed:
Remove unused code

lehecht committed:
Format code

lehecht committed:
Fix indention and format code

lehecht committed:
Merge branch 'main' into pulse-queue-size-card

lehecht committed:
Fix wrong date formatting

mzur commented on README.md:
````diff
@@ -26,7 +26,8 @@ Note: This Laravel package is auto-discovered.
+            'record_interval' => 60, // time interval between records
+            'queues' => ['default'] // queues to monitor
             ],
 ```
````
Please fix the indentation and add the ` ```php ` annotation for syntax highlighting.

mzur reviewed (commented):
When I switch to another time interval (e.g. 6h) the numbers are no longer correct because they are accumulated. Example:

I have a queue with a consistent number of 12 pending jobs. These are recorded at 15:36 to 15:42. When I now switch to the 6h interval, the graph shows a single point "15:42" and accumulates all values of the 6 minutes before that into 72, although it should also show 12. The screenshots show this (but with different numbers):

<img width="1260" height="673" alt="Image" src="https://github.com/user-attachments/assets/f2078023-22da-4064-8501-5359fb5e6988" />

<img width="1260" height="673" alt="Image" src="https://github.com/user-attachments/assets/1ed78b0e-2c89-4242-90ea-8609db3d92c1" />

mzur commented on resources/views/queue-size.blade.php:
```diff
@@ -0,0 +1,173 @@
+                    type: 'line',
+                    data: {
+                        labels: this.labels(config.readings),
+                        datasets: [
+                            {
+                                label: 'Pending',
+                                borderColor: '#9333ea',
+                                data: Object.values(config.readings.pending),
+                            },
+                            {
+                                label: 'Delayed',
+                                borderColor: '#e11d48',
+                                data: Object.values(config.readings.delayed),
+                            },
+                            {
+                                label: 'Reserved',
+                                borderColor: '#eab308',
+                                data: Object.values(config.readings.reserved),
+                            },
+                        ]
```
Also can we please change the colors to:
- pending: red
- delayed: yellow
- reserved: purple

lehecht committed:
Fix php code block in readme

lehecht committed:
Change colors of queue states

lehecht commented:
@mzur 
> When I switch to another time interval (e.g. 6h) the numbers are no longer correct because they are accumulated.

How should the values be aggregated when different values were collected within an interval?

mzur commented:
Take the `max()` of the values.

lehecht committed:
Use max instead of sum

lehecht committed:
Update tests

mzur committed:
Fix indentation

mzur reviewed (approved):
Looks good, thanks!
